WalletWasabi.Backend/Controllers/BlockchainController.cs

```diff
@@ -140,76 +143,83 @@ private async Task<IEnumerable<string>> GetRawMempoolStringsNoCacheAsync(Cancell
+		TaskCompletionSource<Transaction>[] txsCompletionSources = new TaskCompletionSource<Transaction>[requestCount];
+
 		try
 		{
-			var hexes = new Dictionary<uint256, string>();
-			List<uint256> missingTxs = new();
-			lock (TransactionHexCacheLock)
+			// Get task completion sources for transactions. They are either new (no one else is getting that transaction right now) or existing
+			// and then some other caller needs the same transaction so we can use the existing task completion source.
+			for (int i = 0; i < requestCount; i++)
 			{
-				foreach (var txid in parsedIds)
+				uint256 txId = parsedTxIds[i];
+				string cacheKey = $"{nameof(BlockchainController)}#{txId}";
+
+				if (Cache.TryAddKey(cacheKey, TransactionCacheOptions, out TaskCompletionSource<Transaction> tcs))
 				{
-					if (TransactionHexCache.TryGetValue(txid, out string? hex))
-					{
-						hexes.Add(txid, hex);
-					}
-					else
-					{
-						missingTxs.Add(txid);
-					}
+					txIdsRetrieve.Add(txId, tcs);
 				}
+
+				txsCompletionSources[i] = tcs;
 			}
 
-			if (missingTxs.Count != 0)
+			if (txIdsRetrieve.Count > 0)
 			{
-				foreach (var tx in await RpcClient.GetRawTransactionsAsync(missingTxs, cancellationToken))
+				// Ask to get missing transactions over RPC.
+				IEnumerable<Transaction> txs = await RpcClient.GetRawTransactionsAsync(txIdsRetrieve.Keys, cancellationToken).ConfigureAwait(false);
+				Dictionary<uint256, Transaction> rpcBatch = txs.ToDictionary(x => x.GetHash(), x => x);
+
+				foreach (KeyValuePair<uint256, Transaction> kvp in rpcBatch)
 				{
-					string hex = tx.ToHex();
-					hexes.Add(tx.GetHash(), hex);
-
-					lock (TransactionHexCacheLock)
-					{
-						if (TransactionHexCache.TryAdd(tx.GetHash(), hex) && TransactionHexCache.Count >= 1000)
-						{
-							TransactionHexCache.Remove(TransactionHexCache.Keys.First());
-						}
-					}
+					_ = txIdsRetrieve[kvp.Key].TrySetResult(kvp.Value);
 				}
 			}
 
-			// Order hexes according to the order of the query.
-			var orderedResult = parsedIds.Where(x => hexes.ContainsKey(x)).Select(x => hexes[x]);
-			return Ok(orderedResult);
+			Transaction[] result = new Transaction[requestCount];
+
+			// Add missing transactions to the result array.
+			for (int i = 0; i < requestCount; i++)
+			{
+				uint256 txId = parsedTxIds[i];
+				result[i] = await txsCompletionSources[i].Task.ConfigureAwait(false);
+			}
 		}
-		catch (Exception ex)
+		finally
 		{
-			Logger.LogDebug(ex);
-			return BadRequest(ex.Message);
+			if (txIdsRetrieve.Count > 0)
+			{
+				// It's necessary to always set a result to the task completion sources. Otherwise, cache can get corrupted.
+				Exception ex = new InvalidOperationException("Failed to get the transaction.");
+				foreach (TaskCompletionSource<Transaction> tcs in txIdsRetrieve.Values)
+				{
+					_ = tcs.TrySetException(ex);
+				}
+			}
 		}
+
+		return Ok(txsCompletionSources);
```

Comment:
```suggestion
string[] hexes = new string[requestCount];

for (int i = 0; i < requestCount; i++)
{
	hexes[i] = txsCompletionSources[i].Task.Result.ToHex();
}

return Ok(hexes);
```
With this, we don't change the API return type and the response code is 200.